File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Lazy-initialization/Lazy-initialization.java
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Database {

    private static volatile CompletableFuture<Database> connection = null;
    private static final ExecutorService executor = Executors.newSingleThreadExecutor(); 

    private Database() {} 

    // Database connection simulation
    public static synchronized CompletableFuture<String> getConnection() throws InterruptedException {
        if (connection == null) {
            System.out.println("Connecting...");
            connection = CompletableFuture.supplyAsync(() -> {
                try {
                    Thread.sleep(1000); // Simulate connection delay
                    return new Database();
                } catch (Exception e) {
                    throw new RuntimeException(e); 
                }
            }, executor);
        }
        return connection.thenApply(db -> db.getConnectionMessage());
    }

    private String getConnectionMessage() {
        return "Database connection established";
    }

    public static void main(String[] args) throws Exception {
        try {
            CompletableFuture<String> connection1 = Database.getConnection();
            System.out.println(connection1.get()); 

            CompletableFuture<String> connection2 = Database.getConnection();
            System.out.println(connection2.get());
        } catch (Exception e) {
            System.out.println("An error occurred while using the database: " + e.getMessage());
        } finally {
            executor.shutdown(); 
        }
    }
}

File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Lazy-initialization/Lazy-initialization.py
import asyncio
class Database:
    _connection = None
    def __init__(self):
        pass
    # Database connection simulation
    @staticmethod
    async def get_connection():
        if Database._connection is None:
            print("connecting")
            try:
                Database._connection = Database()
            except Exception as error:
              print("Error establishing database connection:", error)
              raise error
        return await Database._connection.get_connection_message()
    async def get_connection_message(self):
        return "Database connection established"
async def main():
  try:
    connection1 = await Database.get_connection()
    print(connection1)

    connection2 = await Database.get_connection()
    print(connection2)
  except Exception as error:
    print("An error occurred while using the database:", error)

asyncio.run(main())